osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs

```diff
@@ -107,9 +114,78 @@ public void AddRotation(float angle)
             }
 
             currentRotation += angle;
+
             // rate has to be applied each frame, because it's not guaranteed to be constant throughout playback
             // (see: ModTimeRamp)
-            drawableSpinner.Result.RateAdjustedRotation += (float)(Math.Abs(angle) * (gameplayClock?.GetTrueGameplayRate() ?? Clock.Rate));
+            double time = gameplayClock?.CurrentTime ?? Clock.CurrentTime;
+            double rate = gameplayClock?.GetTrueGameplayRate() ?? Clock.Rate;
+            float rateAdjustedRotation = (float)(angle * rate);
+
+            if (rate > 0)
+            {
+                int direction = Math.Sign(rateAdjustedRotation);
+
+                if (direction != 0 && direction != currentDirection)
+                {
+                    turningPoints.Add((time, currentDelta));
+                    currentDelta = 0;
+                }
+
+                currentDirection = direction;
+            }
+            else if (turningPoints.Count > 0 && time <= turningPoints[^1].time)
+            {
+                currentDelta = turningPoints[^1].delta;
+                currentDirection = Math.Sign(currentDelta);
+                turningPoints.RemoveAt(turningPoints.Count - 1);
+            }
+
+            currentDelta += rateAdjustedRotation;
+            drawableSpinner.Result.RateAdjustedRotation = computeEffectiveRotation();
+
+            Logger.Log(currentDelta.ToString(CultureInfo.InvariantCulture));
+        }
+
+        private float computeEffectiveRotation()
+        {
+            float curAngle = 0;
+            float maxAngle = 0;
+            float minAngle = 0;
+            float effectiveRotation = 0;
+
+            for (int i = 0; i < turningPoints.Count; i++)
+                applyDelta(turningPoints[i].delta);
+            applyDelta(currentDelta);
+
+            return effectiveRotation;
+
+            void applyDelta(float delta)
+            {
+                float totalBefore = Math.Abs(maxAngle - minAngle);
+
+                curAngle += delta;
+                maxAngle = Math.Max(maxAngle, curAngle);
+                minAngle = Math.Min(minAngle, curAngle);
+
+                float totalAfter = Math.Abs(maxAngle - minAngle);
+
+                effectiveRotation += totalAfter - totalBefore;
+
+                // Check if we've started a new rotation.
+                if (totalAfter < 360)
+                    return;
+
+                // Start a new rotation.
+                // It is fine for totalAfter to skip one rotation (i.e be >= 720), since full rotations are properly accounted for in effectiveRotation.
+                // All that matters is any partial rotation.
+                float excess = (totalAfter % 360) * Math.Sign(delta);
+
+                curAngle = excess;
+                maxAngle = 0;
+                maxAngle = Math.Max(maxAngle, curAngle);
+                minAngle = 0;
+                minAngle = Math.Min(minAngle, curAngle);
```

Comment:
What's going on here? Should this be simplified to 

```suggestion
                maxAngle = Math.Max(0, curAngle);
                minAngle = Math.Min(0, curAngle);
```

Reply:
This is intentional. It resets to 0 and then does the min/max. `curAngle` is an arbitrary value that is either positive or negative. It could be written as:
```
if (curAngle > 0)
{
    maxAngle = curAngle;
    minAngle = 0;
}
else if (curAngle < 0)
{
    maxAngle = 0;
    minAngle = curAngle;
}
else
{
    maxAngle = 0;
    minAngle = 0;
}
```

Reply:
It does not match functionality.

Reply:
Can you double-check? I think there's a misunderstanding somewhere here.

Reply:
You're right, I think I misread something.